[제이미]API 테스트/문서자동화 미션 제출합니다.

jamie9504 opened the pull request:
안녕하세요!!

잘 부탁드립니다 :)

mintjordy committed:
docs: 문서 자동화 실습 요구사항 작성

mintjordy committed:
feat : 회원 API Test 작성 및 Exception, Handler 추가

- 회원 API CRUD Test 작성
- EntityNotFoundException 생성
- Global Exception Handler 생성

mintjordy committed:
feat : 회원 API 문서 자동화(use. Spring Rest Docs)

- 회원 API CRUD에 대한 문서 자동화
- 예외 케이스에 대한 문서 작성 추가

mintjordy committed:
docs : 실습2 요구사항 추가

mintjordy committed:
feat : 인증을 통한 로그인 기능 구현

- Cookie + Session 인증
- Basic(base64) 인증
- Jwt + Bearer 인증

mintjordy committed:
docs : 1단계 회원관리 기능 요구사항 추가

mintjordy committed:
feat : Page URL HTML 화면 연동

- 관리자 페이지 /admin Path Prefix 적용
- 사용자 Index 페이지 연동
- 나의 정보 확인 및 수정 페이지 연동

mintjordy committed:
feat : 로그인 기능 api 테스트 추가

- 세션 기반 로그인시 필요한 정보 parameter -> body로 전달하도록 변경
- 세션 로그인 api 테스트 작성
- 토큰 로그인 api 테스트 작성

mintjordy committed:
feat : 회원 관리 기능 구현

- 사용하지 않는 인증 지원 Intercepter 제거
- 회원 관리 기능 Controller/Service TEST 보강
- 내 정보 조회/수정/삭제 기능 구현
- 중복 Email 회원 가입 예외 처리

mintjordy committed:
docs : 즐겨찾기 요구 사항 및 기능 목록 작성

mintjordy committed:
feat : 회원의 즐겨찾기 전체목록 조회 기능 구현

mintjordy committed:
feat : 회원의 즐겨찾기에 특정 구간이 있는지 확인하는 기능 추가

mintjordy committed:
feat : 회원의 즐겨찾기에 경로 추가 기능 구현

mintjordy committed:
feat : 회원 즐겨찾기 경로 제거 기능 구현

mintjordy committed:
feat : 페이지 연동 기능

kingbbode commented on src/main/java/wooteco/subway/service/favorite/FavoriteService.java:
```diff
@@ -0,0 +1,74 @@
+            memberRepository.save(member);
+        } catch (DbActionExecutionException e) {
+            if (e.getCause() instanceof DuplicateKeyException) {
+                throw new DuplicateEmailException();
```
예외가 잘못 매칭된 것 같아요 :)

jamie9504 replied:
DuplicateEmailException의 경우, 가입 Email이 기존 가입된 Email과 중복되는 경우 발생하는 커스텀 Exception인데요,
 이름이 적합하지 않은 것 같아서 `AlreadyExistsEmailException`로 변경했습니다!

kingbbode commented on src/main/java/wooteco/subway/web/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,40 @@
+import wooteco.subway.web.member.InvalidAuthenticationException;
+
+@RestControllerAdvice(annotations = RestController.class)
+public class GlobalExceptionHandler {
```
예기치 못한 예외도 처리하여, 사용자에게는 정제된 메시지만 돌려주는 것은 어떨까요?

jamie9504 replied:
Exception으로 받아서, 정제된 메시지와 500 코드를 돌려주는 메서드를 만들었습니다!

궁금한 부분은, Spring은 unchecked exception 발생을 지향(?)하는 느낌인데, RuntimeException으로만 잡는 경우도 있는지 궁금합니다!

kingbbode commented on src/test/java/wooteco/subway/service/member/MemberServiceTest.java:
```diff
@@ -1,27 +1,36 @@
+import wooteco.subway.service.member.dto.UpdateMemberRequest;
 
 @ExtendWith(MockitoExtension.class)
 public class MemberServiceTest {
```
구현을 테스트하는 것은 의미가 없습니다 :)

내부 리펙토링 혹은 성능 최적화로 사용하는 메서드와 쿼리가 변경되었을 때 실패하게 될 이 테스트들은 올바른 테스트가 아닙니다.

테스트는 조건을 기반으로 기대하는 결과를 검증하도록 하여, 구현을 모르게 작성하여야 합니다 .

kingbbode replied:
예외에 대한 테스트가 안보이네요 :)

jamie9504 replied:
이번에 verify를 배워서, 아래와 같이 mock 객체의 메서드를 굳이 호출하지 않아도 되는 부분까지 verify를 사용하여 구현에 의존적으로 test를 짜버렸습니다!
``` java
FavoriteRequest request = new FavoriteRequest(FIRST_STATION_ID, THIRD_STATION_ID);
favoriteService.addFavorite(member, request);
verify(memberRepository).save(any());

assertThat(member.getFavorites()).contains(request.toFavorite());
``` 

위와 같은 부분 모두 제거하였습니다.

궁금한 부분이 있는데요,

1. remove와 같은 메서드는 반환값이 void인데, 그럴 경우에도 아래와 같이 테스트 하는 것을 권장하지 않으시는지, 만약 그렇다면 이런 경우엔 어떤 테스트를 해야할지 궁금합니다.
``` java
@Test
void removeLineStationBetweenTwo() {
	when(lineRepository.findById(line.getId())).thenReturn(Optional.of(line));
	lineService.removeLineStation(line.getId(), 2L);

	verify(lineRepository).save(any());
}
```

2. mock을 사용하여 다음과 같이 `when(stationRepository.findAllById(anySet())).thenReturn(stations);`
객체 반환을 미리 지정하는 경우에도 메서드와 쿼리가 변경될 경우 실패하게 되는데, 이 경우에도 올바른 테스트가 아닌지, 아니면 구현을 테스트하는 것이 아닌 구현에 따른 응답값을 셋팅해주는 것이므로 어쩔 수 없이 사용해야 하는 것인지 궁금합니다!

jamie9504 replied:
예외처리 추가했습니다!

아래 부분은 어떻게 Exception을 발생해야하는지 알지 못해서, 제외하고 진행했는데 혹시 어떻게 하면 해당 예외를 발생할 수 있을까요? ㅜ_ㅜ
``` java
catch (DbActionExecutionException e) {
	if (e.getCause() instanceof DuplicateKeyException) {
		throw new AlreadyExistsEmailException();
	}
```

kingbbode commented on src/test/java/wooteco/subway/service/favorite/FavoriteServiceTest.java:
```diff
@@ -0,0 +1,97 @@
+import wooteco.subway.service.station.dto.StationResponse;
+
+@ExtendWith(MockitoExtension.class)
+class FavoriteServiceTest {
```
예외 케이스에 대한 테스트가 보이지 않네요 :)

jamie9504 replied:
Favorite의 경우 위에 남겨놓은 코멘트와 동일한 예외만 서비스에서 처리를 하는데, 혹시 어떻게 발생을 시켜서 테스트를 해봐야할까요? (Mock서버)

``` java
catch (DbActionExecutionException e) {
	if (e.getCause() instanceof DuplicateKeyException) {
		throw new AlreadyExistsEmailException();
	}
```

다음과 같이는 테스트가 되지 않아서요!
``` java
when(memberRepository.save(any())).thenThrow(DuplicateKeyException.class);	when(memberRepository.save(any())).thenThrow(DbActionExecutionException.class);
```

kingbbode commented on build.gradle:
```diff
@@ -12,7 +13,13 @@ repositories {
+}
+
 dependencies {
+	asciidoctor 'org.springframework.restdocs:spring-restdocs-asciidoctor:2.0.4.RELEASE'
+	testImplementation 'org.springframework.restdocs:spring-restdocs-mockmvc:2.0.4.RELEASE'
```
의존성의 버전을 직접 명시해주었네요.
스프링부트에서 관리해주고 있는 버전관리에 대한 부분을 찾아보시면 좋을 듯 합니다 .

jamie9504 replied:
의존성의 명시를 제거해보았습니다.

[Springboot GITHUB](https://github.com/spring-projects/spring-boot/blob/2.2.x/spring-boot-project/spring-boot-dependencies/pom.xml)을 참고해서 제거해보았는데요,

혹시 버전 관리를 인텔리제이를 통해서 조금 더 쉽게 확인할 수 있는 방법이 있는지 궁금합니다!

kingbbode reviewed (changes requested):
안녕하세요 :) 잘 구현해주셨네요.
몇 가지 코멘트 추가해두었습니다!
확인 부탁드려요.

jamie9504 committed:
feat : 예기치 못한 예외 핸들링

jamie9504 committed:
feat : Body 컨트롤러 검증 및 핸들러 추가

- MethodArgumentNotValidException 핸들러 추가 및 예외 메시지 구현
- @Validated 애너테이션 검증 로직 추가

jamie9504 committed:
refactor: 토큰 헤더값 Prefix Enum으로 관리

jamie9504 committed:
refactor: 미사용 코드 제거 및 Generic 타입 추가

- 로 타입 제네릭 타입으로 변경(컨트롤러)
- 미사용 테스트 코드 제거

jamie9504 committed:
refactor: DB FK 지정 및 data 쿼리 개선

- DB FK 지정(삭제시 같이 삭제되게끔)
- data 쿼리 insert values 통합
- 성능을 위해 모든 DB ID 넣어줌

jamie9504 committed:
refactor: Exception명 변경 및 코드 정리(컨벤션)

- Exception : AlreadyExistsEmailException

jamie9504 committed:
refactor: @Transactional, 패키지 이동, PK 추가

- 서비스에 @Transactional 애너테이션 붙임
- ErrorResponse 패키지 위치 이동: controller에서 쓰이기 때문에 controller 패키지 하위로 이동
- ID에 PK 추가

jamie9504 committed:
refactor: withId 메서드 추가 및 빈생성자 제거

- withId 메서드 추가
- 빈 생성자 제거
- 전체 생성자 private 설정
- 생성자 -> 정적 팩터리 메서드 전환

jamie9504 committed:
refactor : entity 불변으로 변경

- entity 불변으로 변경(성능)

jamie9504 committed:
refactor : verify 테스트 assertThat으로 변경

- 구현 내용을 확인하는 테스트가 아닌 결과 확인 테스트로 변경

jamie9504 committed:
refactor : MemberServiceTest 예외 테스트 추가

jamie9504 committed:
refactor : build.gradle 의존성 중복 관리 제거

kingbbode reviewed (approved):
고생 하셨어요! 이만 머지하겠습니다 :)
